controllers/volumereplicationgroup_controller.go

```diff
@@ -1029,72 +1158,292 @@ func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName, state volr
-		return v.validateVRStatusAsPrimary(volRep)
-	case v.instance.Spec.ReplicationState == volrep.Secondary:
-		return v.validateVRStatusAsSecondary(volRep)
-	case v.instance.Spec.ReplicationState == volrep.Resync:
-		return fmt.Errorf("resync is not supported for VRG")
+	case v.instance.Spec.ReplicationState == ramendrv1alpha1.Primary:
+		return v.validateVRStatus(volRep, ramendrv1alpha1.Primary)
+	case v.instance.Spec.ReplicationState == ramendrv1alpha1.Secondary:
+		return v.validateVRStatus(volRep, ramendrv1alpha1.Secondary)
 	default:
+		msg := "VolumeReplicationGroup state invalid"
+		v.updateProtectedPVCCondition(volRep.Name, PVCError, msg)
+
 		return fmt.Errorf("invalid Replication State %s for VolumeReplicationGroup (%s:%s)",
 			string(v.instance.Spec.ReplicationState), v.instance.Name, v.instance.Namespace)
 	}
 }
 
-// So, VolumeReplication controller reflects the status of a VolRep resource as primary
-// only after the promotion of the associated storage volume to primary is successful.
-// Hence, checking whether the primary state is reflected in the status is sufficient
-// to verify whether the VolRep resource is in a healthy state not. However, checking
-// conditions of VolRep status may be required for maintaining correct status of VRG
-// when VRG status is implemented.
+// VolumeReplication controller sets volRep.Status.State to primary/secondary
+// only after the successful promotion/demotion of the underlying storage volume
+// consumed by the pvc.It is necessary to get the appropriate condition out of
+// different conditions a VolRep status can have. Doing this may be helpful in
+// getting a more clear view of the VolumeReplication resource's current situation.
 //
-// TODO: If checking conditions of VolRep is necessary for VRG status, add that logic
-//       here when VRG status is implemented.
-func (v *VRGInstance) validateVRStatusAsPrimary(volRep *volrep.VolumeReplication) error {
-	if volRep.Status.State != volrep.PrimaryState {
-		v.log.Info("State is primary and status is not primary")
+// As of now, name of the VolumeReplication resource and pvc that it represents is
+// same. So, using the name from volRep resource to get protectedPVC condition from
+// vrg.Status. In future if things change, some changes might have to be necessary.
+func (v *VRGInstance) validateVRStatus(volRep *volrep.VolumeReplication, state ramendrv1alpha1.ReplicationState) error {
+	var (
+		stateString = "unknown"
+		action      = "unknown"
+	)
 
-		return fmt.Errorf("primary VolumeReplication resource status not same %s/%s (status: %s)",
-			volRep.Name, volRep.Namespace, volRep.Status.State)
+	switch state {
+	case ramendrv1alpha1.Primary:
+		stateString = "primary"
+		action = "promoted"
+	case ramendrv1alpha1.Secondary:
+		stateString = "secondary"
+		action = "demoted"
 	}
 
-	v.log.Info("State in the spec and status match for primary")
+	volRepCondition := findCondition(volRep.Status.Conditions, volrepController.ConditionCompleted)
```

Comment:
We need to ensure condition generation is the same as VR resource generation, so that we are not processing stale/older conditions.

Although we should not hit this, the issue is if condition we are checking is true for an older generation and we just changed the spec. So we need to validate the condition generation first and then if that matches, check the conditions that we want.

Reply:
So, the function checkVRStatus() actually compares the generation number for the VolRep resource and VolRep.Status. If they are not same, it would return instead of proceeding to check the conditions. I thought that would be sufficient enough. Hence, by the time the code arrives here, that would mean the genration number is same. 

If you think that comparing the generation number of VolRep resource and its status is not sufficient and an explicit check has to be made (either in addition to or instead of the check I mentioned in above paragraph) for conditions.ObservedGeneration as well, let me know.

Reply:
Ack!